src/main/java/net/neoforged/neoforge/capabilities/CapabilityHooks.java

```diff
@@ -0,0 +1,170 @@
+            // Return a wrapper that gets re-evaluated every time it is accessed
+            // Invalidation is taken care of by the patches to ComposterBlock
+            if (side == null)
+                return new ForwardingItemHandler(() -> new InvWrapper(composterBlock.getContainer(state, level, pos)));
+            else
+                return new ForwardingItemHandler(() -> new SidedInvWrapper(composterBlock.getContainer(state, level, pos), side));
+        }, Blocks.COMPOSTER);
+
+        event.registerBlock(Capabilities.ItemHandler.BLOCK, (level, pos, state, blockEntity, side) -> {
+            return new InvWrapper(ChestBlock.getContainer((ChestBlock) state.getBlock(), state, level, pos, true));
+        }, Blocks.CHEST, Blocks.TRAPPED_CHEST);
+
+        event.registerBlockEntity(Capabilities.ItemHandler.BLOCK, BlockEntityType.HOPPER, (hopper, side) -> {
+            // Use custom hopper wrapper that respects cooldown
+            return new VanillaHopperItemHandler(hopper);
+        });
+
+        event.registerBlockEntity(Capabilities.ItemHandler.BLOCK, BlockEntityType.SHULKER_BOX, (shulkerBox, side) -> {
+            // Always use UP wrapper for shulker boxes
+            return new SidedInvWrapper(shulkerBox, Direction.UP);
+        });
+
+        var sidedVanillaContainers = List.of(
+                BlockEntityType.BLAST_FURNACE,
+                BlockEntityType.BREWING_STAND,
+                BlockEntityType.FURNACE,
+                BlockEntityType.SMOKER);
+        for (var bet : sidedVanillaContainers) {
+            event.registerBlockEntity(Capabilities.ItemHandler.BLOCK, bet, (sidedContainer, side) -> {
+                return side == null ? new InvWrapper(sidedContainer) : new SidedInvWrapper(sidedContainer, side);
+            });
+        }
+
+        var unsidedVanillaContainers = List.of(
+                BlockEntityType.BARREL,
+                BlockEntityType.CHISELED_BOOKSHELF,
+                BlockEntityType.DISPENSER,
+                BlockEntityType.DROPPER,
+                BlockEntityType.JUKEBOX);
+        for (var bet : unsidedVanillaContainers) {
+            event.registerBlockEntity(Capabilities.ItemHandler.BLOCK, bet, (container, side) -> {
+                return new InvWrapper(container);
+            });
+        }
+
+        // Entities
+        var containerEntities = List.of(
+                EntityType.CHEST_BOAT,
+                EntityType.CHEST_MINECART,
+                EntityType.HOPPER_MINECART);
+        for (var entityType : containerEntities) {
+            event.registerEntity(Capabilities.ItemHandler.ENTITY, entityType, (entity, ctx) -> new InvWrapper(entity));
+            event.registerEntity(Capabilities.ItemHandler.ENTITY_AUTOMATION, entityType, (entity, ctx) -> new InvWrapper(entity));
+        }
+        event.registerEntity(Capabilities.ItemHandler.ENTITY, EntityType.PLAYER, (player, ctx) -> new PlayerInvWrapper(player.getInventory()));
```

Comment:
This is ever so slightly off topic of this PR, but is it possible to make the `PlayerInvWrapper` share a single `InvWrapper` between the three combined wrappers rather than have each one create their own wrapper?

Reply:
This is a bit tricky due to the `insertItem` overloads. I don't think the extra allocations really matter TBH. Worst case we can store the wrapper as a field in the player if we see overhead coming from there.